contracts/ronin/BridgeTracking.sol

```diff
@@ -190,42 +203,55 @@ contract BridgeTracking is HasBridgeContract, HasValidatorContract, Initializabl
     address _bridgeOperator,
     bool _mustCountLastStats
   ) internal view returns (uint256 _totalBallots) {
-    _totalBallots = _periodStats[_period].totalBallotsOf[_bridgeOperator];
+    _totalBallots = _periodMetric[_period].totalBallotsOf[_bridgeOperator];
     if (_mustCountLastStats) {
-      _totalBallots += _temporaryStats.info.totalBallotsOf[_bridgeOperator];
+      _totalBallots += _bufferMetric.data.totalBallotsOf[_bridgeOperator];
     }
   }
 
   /**
-   * @dev Syncs period stats if the last epoch + 1 is already wrapped up.
+   * @dev Syncs period stats. Move all data from the buffer metric to a new period metric.
+   *
+   * Requirements:
+   * - The buffer epoch + 1 is already wrapped up.
    */
-  function _trySyncPeriodStats() internal {
+  function _trySyncBuffer() internal {
     uint256 _currentEpoch = _validatorContract.epochOf(block.number);
-    if (_temporaryStats.lastEpoch < _currentEpoch) {
-      (, uint256 _period) = _validatorContract.tryGetPeriodOfEpoch(_temporaryStats.lastEpoch + 1);
-      VoteStats storage _stats = _periodStats[_period];
-      _stats.totalVotes += _temporaryStats.info.totalVotes;
-      _stats.totalBallots += _temporaryStats.info.totalBallots;
-
-      address _voter;
-      for (uint _i = 0; _i < _temporaryStats.info.voters.length; _i++) {
-        _voter = _temporaryStats.info.voters[_i];
-        _stats.totalBallotsOf[_voter] += _temporaryStats.info.totalBallotsOf[_voter];
-        delete _temporaryStats.info.totalBallotsOf[_voter];
+    if (_bufferMetric.lastEpoch < _currentEpoch) {
+      (, uint256 _currentPeriod) = _validatorContract.tryGetPeriodOfEpoch(_bufferMetric.lastEpoch + 1);
+      _bufferMetric.lastEpoch = _currentEpoch;
+
+      PeriodVotingMetric storage _metric = _periodMetric[_currentPeriod];
+      _metric.totalBallots += _bufferMetric.data.totalBallots;
+
+      for (uint _i = 0; _i < _bufferMetric.data.voters.length; _i++) {
+        address _voter = _bufferMetric.data.voters[_i];
+        _metric.totalBallotsOf[_voter] += _bufferMetric.data.totalBallotsOf[_voter];
+        delete _bufferMetric.data.totalBallotsOf[_voter]; // need to manually delete each element, due to mapping
+      }
+
+      for (uint _i = 0; _i < _bufferMetric.data.requests.length; _i++) {
+        Request storage _bufferRequest = _bufferMetric.data.requests[_i];
+        Request storage _metricRequest = _metric.requests.push();
+        _metricRequest.kind = _bufferRequest.kind;
+        _metricRequest.id = _bufferRequest.id;
```

Comment:
I cannot see any function to expose the request array in `_periodMetric` (except the length). Should it be exposed somewhere or should we store the length instead?

Reply:
The list of requests must be recorded in each period metric, in order to update the `trackedPeriod` for each request later.

https://github.com/axieinfinity/ronin-dpos-contracts/blob/43db61a8e9e14224971556318e7993846a9c97b5/contracts/ronin/BridgeTracking.sol#L239-L240